File: packages/routing-engine/src/rules.ts
/**
 * Routing rule schema — the declarative contract an admin edits and both engines execute.
 *
 * The schema is deliberately data-only: no callbacks, no embedded code. That is what lets
 * a rule set be authored in the admin UI, signed into a bundle, shipped to 20-30 agents and
 * executed identically by the C# engine on the workstation and the TypeScript engine on the
 * server. Anything a rule cannot express has to become a new predicate here first, in both
 * implementations, with a conformance fixture — never a special case in one engine.
 */

/** Inclusive numeric bound. Omitting a side leaves it unbounded. */
export interface RangeMm {
  min?: number;
  max?: number;
}

/**
 * A rectangle a predicate or transform operates on.
 *
 * `inkBox` and `barcodeCluster` are resolved per page from the extracted features, which is
 * what makes one rule work across the 25+ document shapes in the corpus: the label region is
 * found by measurement, not by a hard-coded position.
 */
export type RectSpec =
  | 'page'
  | 'inkBox'
  | 'barcodeCluster'
  | { unit: 'mm'; xMm: number; yMm: number; widthMm: number; heightMm: number }
  | { unit: 'pageFraction'; x: number; y: number; w: number; h: number };

/** Matches against the embedded PDF text layer. */
export interface TextPredicate {
  contains?: string;
  /** JavaScript-flavoured regular expression source; `.NET` accepts the same subset. */
  matches?: string;
  /** Restrict the match to text inside this rectangle. Requires positioned text. */
  withinRect?: RectSpec;
  caseSensitive?: boolean;
}

/**
 * Matches against OCR output, forcing OCR of `rect` when the host has not supplied it.
 *
 * OCR text is normalised before matching (runs of whitespace collapse to a single space)
 * because recognisers routinely drop the spaces in a bold header: the DHL sheet comes back
 * as `*WAYBILLDOC*`, so a rule must be able to write `WAYBILL\s*DOC` and mean it.
 */
export interface OcrPredicate {
  rect: RectSpec;
  contains?: string;
  matches?: string;
  caseSensitive?: boolean;
  /** Also try the match with all whitespace removed. Defaults to true. */
  ignoreSpacing?: boolean;
}

/** Matches decoded barcodes. */
export interface BarcodePredicate {
  /** zxing symbology names, e.g. `Code128`, `PDF417`, `MaxiCode`, `DataMatrix`. */
  symbology?: string[];
  valueMatches?: string;
  valueContains?: string;
  minCount?: number;
  maxCount?: number;
  /** Only count barcodes whose box lies inside this rectangle. */
  rect?: RectSpec;
}

/** Print&Share-style picture matching against a named template in the bundle. */
export interface ImagePredicate {
  template: string;
  /** Minimum normalised correlation score, 0..1. */
  threshold: number;
  searchRect?: RectSpec;
}

/** Matches measured page and ink geometry. All bounds are millimetres. */
export interface GeometryPredicate {
  orientation?: 'portrait' | 'landscape';
  pageWidthMm?: RangeMm;
  pageHeightMm?: RangeMm;
  inkWidthMm?: RangeMm;
  inkHeightMm?: RangeMm;
  inkXMm?: RangeMm;
  inkYMm?: RangeMm;
  /** Ink height / ink width. A 4x6in label is 1.5, the DHL label ~1.97. */
  inkAspect?: RangeMm;
  /** Fraction of the page covered in ink, 0..1. */
  inkCoverage?: RangeMm;
  /** True when the page itself is label stock rather than a sheet carrying a label. */
  pageIsLabelStock?: boolean;
}

/** Matches the resolved carrier. */
export interface CarrierPredicate {
  is?: string;
  in?: string[];
  /** Minimum carrier-resolution confidence, 0..1. */
  minConfidence?: number;
}

/** Matches the page position inside the document. */
export interface PageIndexPredicate {
  is?: 'first' | 'last';
  /** 1-based. */
  nth?: number;
  range?: { min?: number; max?: number };
}

/** A rule condition. Composite forms nest arbitrarily. */
export type Predicate =
  | { all: Predicate[] }
  | { any: Predicate[] }
  | { not: Predicate }
  | { text: TextPredicate }
  | { ocr: OcrPredicate }
  | { barcode: BarcodePredicate }
  | { image: ImagePredicate }
  | { geometry: GeometryPredicate }
  | { carrier: CarrierPredicate }
  | { pageIndex: PageIndexPredicate };

export type RotateSpec = 'auto' | 0 | 90 | 180 | 270;
export type FitSpec = 'contain' | 'cover' | 'actual' | 'stretch';

/**
 * How the chosen region is placed on the target media.
 *
 * `media` is a free `WxH mm` string (or a named size) rather than an enum, so 100x150,
 * 100x200 and 105x148 all work with no code change — see plan section 7.4.
 */
export interface TransformSpec {
  /** Region of the page to print. Defaults to the whole page. */
  source?: RectSpec;
  /** Grow the source region by this much on every side before fitting. */
  padMm?: number;
  rotate?: RotateSpec;
  fit?: FitSpec;
  /** `100x150mm`, `100x150`, `A4`, ... */
  media?: string;
  zoomPercent?: number;
  panXMm?: number;
  panYMm?: number;
  copies?: number;
  colorMode?: 'mono' | 'color';
  duplex?: 'simplex' | 'long-edge' | 'short-edge';
  tray?: string;
}

/** What happens when a rule matches. */
export interface RuleAction {
  /** Role (`A4`, `THERMAL`) or a named printer alias from the printer map. */
  route?: string;
  transform?: TransformSpec;
  copies?: number;
  /** Queue the page for confirmation instead of printing it. */
  hold?: boolean;
  /** Confidence this rule asserts when it matches, 0..1. Defaults to 1. */
  confidence?: number;
  /** When false, evaluation continues to later rules. Defaults to true. */
  stop?: boolean;
}

export interface PageRule {
  /** Stable identifier, referenced by traces and by the review queue. */
  id: string;
  name: string;
  when: Predicate;
  then: RuleAction;
  /** Disabled rules stay in the bundle and in the trace, but never match. */
  enabled?: boolean;
}

/** Which documents a profile applies to. */
export interface ProfileMatch {
  /** Glob against the source file or job name, e.g. `OneClickPrint_*.pdf`. */
  filenameMask?: string;
  sourceApp?: string;
  minPages?: number;
  maxPages?: number;
}

export type FallbackBehaviour = 'prompt' | 'route' | 'hold';

/** Reason a page could not be routed with confidence. See plan section 6.6. */
export type FallbackReason =
  | 'NO_THERMAL_CANDIDATE'
  | 'LOW_CONFIDENCE'
  | 'AMBIGUOUS'
  | 'UNKNOWN_CARRIER'
  | 'NO_PROFILE_MATCH'
  | 'SERVER_UNAVAILABLE'
  | 'RULE_HOLD'
  | 'CROP_IMPLAUSIBLE'
  | 'RENDER_FAILED'
  | 'DECODE_FAILED';

export interface FallbackPolicy {
  /** Route used when nothing matched and the behaviour is `route`. */
  route: string;
  /** Default behaviour for any reason not named in `byReason`. */
  onUnknown: FallbackBehaviour;
  byReason?: Partial<Record<FallbackReason, FallbackBehaviour>>;
}

/**
 * What the profile expects a matching document to contain, so the engine can tell
 * "no label found" from "this document legitimately has no label".
 */
export interface DocumentExpectations {
  thermalPagesPerDocument?: RangeMm;
}

export interface RoutingProfileRules {
  /** Human-facing profile name, e.g. `Marendo OneClickPrint`. */
  profile: string;
  /** Bundle version this rule set was published as. */
  version?: number;
  match?: ProfileMatch;
  /** Pages below this confidence escalate or fall back. Defaults to 0.75. */
  confidenceThreshold?: number;
  pageRules: PageRule[];
  fallback: FallbackPolicy;
  expectations?: DocumentExpectations;
}

export const DEFAULT_CONFIDENCE_THRESHOLD = 0.75;

/** Roles every deployment has; anything else is a printer alias. */
export const ROUTE_A4 = 'A4';
export const ROUTE_THERMAL = 'THERMAL';

File: clients/windows/Printo.Agent.Core/Routing/Rules.cs
using System.Text.Json.Serialization;

namespace Printo.Agent.Core.Routing;

/// <summary>
/// Routing rule schema — the declarative contract an admin edits and both engines execute.
/// </summary>
/// <remarks>
/// Mirrors <c>packages/routing-engine/src/rules.ts</c>. The schema is deliberately data-only:
/// no callbacks, no embedded code. That is what lets a rule set be authored in the admin UI,
/// signed into a bundle, shipped to the fleet and executed identically here and on the server.
/// Anything a rule cannot express has to become a new predicate in both implementations, with
/// a conformance fixture — never a special case in one engine.
/// </remarks>
public sealed class RangeMm
{
    public double? Min { get; init; }

    public double? Max { get; init; }

    public bool Contains(double value) =>
        (Min is null || value >= Min.Value) && (Max is null || value <= Max.Value);

    public override string ToString() => (Min, Max) switch
    {
        (not null, not null) => $"{Min}..{Max}",
        (not null, null) => $">={Min}",
        (null, not null) => $"<={Max}",
        _ => "any",
    };
}

/// <summary>How a rectangle is named in a rule. Serialized as a string or an object.</summary>
[JsonConverter(typeof(RectSpecConverter))]
public sealed class RectSpec
{
    /// <summary><c>page</c>, <c>inkBox</c> or <c>barcodeCluster</c>; null for an explicit rect.</summary>
    public string? Named { get; init; }

    /// <summary><c>mm</c> or <c>pageFraction</c>; null when <see cref="Named"/> is set.</summary>
    public string? Unit { get; init; }

    public double X { get; init; }

    public double Y { get; init; }

    public double W { get; init; }

    public double H { get; init; }

    public static RectSpec Page { get; } = new() { Named = "page" };

    public static RectSpec InkBox { get; } = new() { Named = "inkBox" };

    public override string ToString() => Named ?? (Unit == "mm"
        ? $"{X},{Y} {W}x{H}mm"
        : $"{X},{Y} {W}x{H} of page");
}

public abstract class Predicate;

public sealed class AllPredicate : Predicate
{
    public IReadOnlyList<Predicate> All { get; init; } = [];
}

public sealed class AnyPredicate : Predicate
{
    public IReadOnlyList<Predicate> Any { get; init; } = [];
}

public sealed class NotPredicate : Predicate
{
    public Predicate Not { get; init; } = new GeometryPredicate { Geometry = new GeometryCondition() };
}

/// <summary>Matches against the embedded PDF text layer.</summary>
public sealed class TextCondition
{
    public string? Contains { get; init; }

    /// <summary>Regular expression source. The .NET and JavaScript engines accept the same subset.</summary>
    public string? Matches { get; init; }

    /// <summary>Restrict the match to text inside this rectangle. Requires positioned text.</summary>
    public RectSpec? WithinRect { get; init; }

    public bool? CaseSensitive { get; init; }
}

public sealed class TextPredicate : Predicate
{
    public TextCondition Text { get; init; } = new();
}

/// <summary>
/// Matches against OCR output, forcing OCR of <see cref="Rect"/> when the host has not
/// supplied it. Matching is whitespace-tolerant because recognisers routinely drop the
/// spaces in a bold header: the DHL courier sheet comes back as <c>*WAYBILLDOC*</c>.
/// </summary>
public sealed class OcrCondition
{
    public RectSpec Rect { get; init; } = RectSpec.InkBox;

    public string? Contains { get; init; }

    public string? Matches { get; init; }

    public bool? CaseSensitive { get; init; }

    /// <summary>Also try the match with all whitespace removed. Defaults to true.</summary>
    public bool? IgnoreSpacing { get; init; }
}

public sealed class OcrPredicate : Predicate
{
    public OcrCondition Ocr { get; init; } = new();
}

public sealed class BarcodeCondition
{
    public IReadOnlyList<string>? Symbology { get; init; }

    public string? ValueMatches { get; init; }

    public string? ValueContains { get; init; }

    public int? MinCount { get; init; }

    public int? MaxCount { get; init; }

    public RectSpec? Rect { get; init; }
}

public sealed class BarcodePredicate : Predicate
{
    public BarcodeCondition Barcode { get; init; } = new();
}

public sealed class ImageCondition
{
    public string Template { get; init; } = string.Empty;

    /// <summary>Minimum normalised correlation score, 0..1.</summary>
    public double Threshold { get; init; }

    public RectSpec? SearchRect { get; init; }
}

public sealed class ImagePredicate : Predicate
{
    public ImageCondition Image { get; init; } = new();
}

/// <summary>Matches measured page and ink geometry. All bounds are millimetres.</summary>
public sealed class GeometryCondition
{
    public PageOrientation? Orientation { get; init; }

    public RangeMm? PageWidthMm { get; init; }

    public RangeMm? PageHeightMm { get; init; }

    public RangeMm? InkWidthMm { get; init; }

    public RangeMm? InkHeightMm { get; init; }

    public RangeMm? InkXMm { get; init; }

    public RangeMm? InkYMm { get; init; }

    /// <summary>Ink height / ink width. A 4x6in label is 1.5, the DHL label ~1.97.</summary>
    public RangeMm? InkAspect { get; init; }

    /// <summary>Fraction of the page covered in ink, 0..1.</summary>
    public RangeMm? InkCoverage { get; init; }

    /// <summary>True when the page itself is label stock rather than a sheet carrying a label.</summary>
    public bool? PageIsLabelStock { get; init; }
}

public sealed class GeometryPredicate : Predicate
{
    public GeometryCondition Geometry { get; init; } = new();
}

public sealed class CarrierCondition
{
    public string? Is { get; init; }

    public IReadOnlyList<string>? In { get; init; }

    public double? MinConfidence { get; init; }
}

public sealed class CarrierPredicate : Predicate
{
    public CarrierCondition Carrier { get; init; } = new();
}

public sealed class PageIndexCondition
{
    /// <summary><c>first</c> or <c>last</c>.</summary>
    public string? Is { get; init; }

    /// <summary>1-based.</summary>
    public int? Nth { get; init; }

    public RangeMm? Range { get; init; }
}

public sealed class PageIndexPredicate : Predicate
{
    public PageIndexCondition PageIndex { get; init; } = new();
}

/// <summary>
/// How the chosen region is placed on the target media. <see cref="Media"/> is a free
/// <c>WxH mm</c> string rather than an enum, so any stock works without a code change.
/// </summary>
public sealed class TransformSpec
{
    public RectSpec? Source { get; init; }

    public double? PadMm { get; init; }

    /// <summary><c>auto</c>, <c>0</c>, <c>90</c>, <c>180</c> or <c>270</c>.</summary>
    [JsonConverter(typeof(RotateSpecConverter))]
    public RotateSpec? Rotate { get; init; }

    /// <summary><c>contain</c> (default), <c>cover</c>, <c>actual</c> or <c>stretch</c>.</summary>
    public string? Fit { get; init; }

    public string? Media { get; init; }

    public double? ZoomPercent { get; init; }

    public double? PanXMm { get; init; }

    public double? PanYMm { get; init; }

    public int? Copies { get; init; }

    public string? ColorMode { get; init; }

    public string? Duplex { get; init; }

    public string? Tray { get; init; }
}

/// <summary><c>auto</c> or an explicit quarter turn.</summary>
public sealed class RotateSpec
{
    public bool IsAuto { get; init; }

    public int Degrees { get; init; }

    public static RotateSpec Auto { get; } = new() { IsAuto = true };

    public static RotateSpec Fixed(int degrees) => new() { Degrees = degrees };
}

public sealed class RuleAction
{
    /// <summary>Role (<c>A4</c>, <c>THERMAL</c>) or a named printer alias.</summary>
    public string? Route { get; init; }

    public TransformSpec? Transform { get; init; }

    public int? Copies { get; init; }

    /// <summary>Queue the page for confirmation instead of printing it.</summary>
    public bool? Hold { get; init; }

    /// <summary>Confidence this rule asserts when it matches, 0..1. Defaults to 1.</summary>
    public double? Confidence { get; init; }

    /// <summary>When false, evaluation continues to later rules. Defaults to true.</summary>
    public bool? Stop { get; init; }
}

public sealed class PageRule
{
    public string Id { get; init; } = string.Empty;

    public string Name { get; init; } = string.Empty;

    public Predicate When { get; init; } = new GeometryPredicate();

    public RuleAction Then { get; init; } = new();

    public bool? Enabled { get; init; }
}

public sealed class ProfileMatch
{
    /// <summary>Glob against the source file or job name, e.g. <c>OneClickPrint_*.pdf</c>.</summary>
    public string? FilenameMask { get; init; }

    public string? SourceApp { get; init; }

    public int? MinPages { get; init; }

    public int? MaxPages { get; init; }
}

/// <summary>Reason a page could not be routed with confidence.</summary>
public enum FallbackReason
{
    [JsonStringEnumMemberName("NO_THERMAL_CANDIDATE")]
    NoThermalCandidate,

    [JsonStringEnumMemberName("LOW_CONFIDENCE")]
    LowConfidence,

    [JsonStringEnumMemberName("AMBIGUOUS")]
    Ambiguous,

    [JsonStringEnumMemberName("UNKNOWN_CARRIER")]
    UnknownCarrier,

    [JsonStringEnumMemberName("NO_PROFILE_MATCH")]
    NoProfileMatch,

    [JsonStringEnumMemberName("SERVER_UNAVAILABLE")]
    ServerUnavailable,

    [JsonStringEnumMemberName("RULE_HOLD")]
    RuleHold,

    [JsonStringEnumMemberName("CROP_IMPLAUSIBLE")]
    CropImplausible,

    [JsonStringEnumMemberName("RENDER_FAILED")]
    RenderFailed,

    [JsonStringEnumMemberName("DECODE_FAILED")]
    DecodeFailed,
}

/// <summary>Wire names for <see cref="FallbackReason"/>, shared with the TypeScript engine.</summary>
public static class FallbackReasons
{
    public static string ToWire(FallbackReason reason) => reason switch
    {
        FallbackReason.NoThermalCandidate => "NO_THERMAL_CANDIDATE",
        FallbackReason.LowConfidence => "LOW_CONFIDENCE",
        FallbackReason.Ambiguous => "AMBIGUOUS",
        FallbackReason.UnknownCarrier => "UNKNOWN_CARRIER",
        FallbackReason.NoProfileMatch => "NO_PROFILE_MATCH",
        FallbackReason.ServerUnavailable => "SERVER_UNAVAILABLE",
        FallbackReason.RuleHold => "RULE_HOLD",
        FallbackReason.CropImplausible => "CROP_IMPLAUSIBLE",
        FallbackReason.RenderFailed => "RENDER_FAILED",
        FallbackReason.DecodeFailed => "DECODE_FAILED",
        _ => reason.ToString(),
    };

    public static FallbackReason FromWire(string wire) => wire switch
    {
        "NO_THERMAL_CANDIDATE" => FallbackReason.NoThermalCandidate,
        "LOW_CONFIDENCE" => FallbackReason.LowConfidence,
        "AMBIGUOUS" => FallbackReason.Ambiguous,
        "UNKNOWN_CARRIER" => FallbackReason.UnknownCarrier,
        "NO_PROFILE_MATCH" => FallbackReason.NoProfileMatch,
        "SERVER_UNAVAILABLE" => FallbackReason.ServerUnavailable,
        "RULE_HOLD" => FallbackReason.RuleHold,
        "CROP_IMPLAUSIBLE" => FallbackReason.CropImplausible,
        "RENDER_FAILED" => FallbackReason.RenderFailed,
        "DECODE_FAILED" => FallbackReason.DecodeFailed,
        _ => throw new ArgumentOutOfRangeException(nameof(wire), wire, "Unknown fallback reason"),
    };
}

public enum FallbackBehaviour
{
    [JsonStringEnumMemberName("prompt")]
    Prompt,

    [JsonStringEnumMemberName("route")]
    Route,

    [JsonStringEnumMemberName("hold")]
    Hold,
}

public sealed class FallbackPolicy
{
    /// <summary>Route used when nothing matched and the behaviour is <c>route</c>.</summary>
    public string Route { get; init; } = RoutingProfileRules.RouteA4;

    /// <summary>Default behaviour for any reason not named in <see cref="ByReason"/>.</summary>
    public FallbackBehaviour OnUnknown { get; init; } = FallbackBehaviour.Prompt;

    /// <summary>
    /// Keyed by the reason's wire name (<c>LOW_CONFIDENCE</c>, ...) rather than by the enum:
    /// dictionary keys do not go through the string-enum converter, so using the enum here
    /// would silently serialize as integers and break bundle interchange with the server.
    /// </summary>
    public IReadOnlyDictionary<string, FallbackBehaviour>? ByReason { get; init; }

    /// <summary>Behaviour configured for a reason, or <see cref="OnUnknown"/>.</summary>
    public FallbackBehaviour For(FallbackReason reason) =>
        ByReason is not null && ByReason.TryGetValue(FallbackReasons.ToWire(reason), out var behaviour)
            ? behaviour
            : OnUnknown;
}

/// <summary>
/// What the profile expects a matching document to contain, so the engine can tell
/// "no label found" from "this document legitimately has no label".
/// </summary>
public sealed class DocumentExpectations
{
    public RangeMm? ThermalPagesPerDocument { get; init; }
}

public sealed class RoutingProfileRules
{
    /// <summary>Roles every deployment has; anything else is a printer alias.</summary>
    public const string RouteA4 = "A4";

    public const string RouteThermal = "THERMAL";

    public const double DefaultConfidenceThreshold = 0.75;

    public string Profile { get; init; } = string.Empty;

    public int? Version { get; init; }

    public ProfileMatch? Match { get; init; }

    /// <summary>Pages below this confidence escalate or fall back. Defaults to 0.75.</summary>
    public double? ConfidenceThreshold { get; init; }

    public IReadOnlyList<PageRule> PageRules { get; init; } = [];

    public FallbackPolicy Fallback { get; init; } = new();

    public DocumentExpectations? Expectations { get; init; }
}
